### ISSUE TITLE
Enhance Tessel port detection logic

### ISSUE THREAD

**Issue by wyze** [tessel/ambient-attx4#63]
As it is currently, this is what is used:

```js
function use(hardware, callback) {
  if (!hardware) {
    console.warn(new Error('Improperly specified Tessel port.'));
  }

  return new Ambient(hardware, callback);
}
```

1. Why `console.warn` instead of `throw`? If it does warn, it is going to throw somewhere else, we should give feedback as early as possible.

2. Can this check be moved into the `Ambient` function to also catch rare cases where someone is not using the `use` function and just `Ambient` directly?

3. I think the logic needs to change to be a little more robust. Right now, it just checks to see if the value is truthy, so it won't go into the `if` statement if say I do something like `var ambient = ambientlib.use(true);`. I think if tessel/t2-firmware#172 could land we can do a lot better here.

Example: 
```js
  if (
    typeof hardware !== 'object' ||
    (typeof hardware === 'object' && hardware.constructor.name !== 'TesselPort')
  ) {
    throw new Error('Improperly specified Tessel port.');
  }
```

**Comment by wyze** [comment]
It sounds like `3.` isn't a good solution. Given that, what would be a better way to accomplish the end goal here? I think making sure `hardware` param is an object instead of just `truthy` is good. Maybe check that a certain property exists?

Updated Example:

```js
  if (typeof hardware !== 'object' || !hardware.board) {
    throw new Error('Improperly specified Tessel port.');
  }
```

**Comment by rwaldron** [comment]
If you add the following to their respective libraries:

```js
// In t2-firmware/node/tessel-export.js:
Tessel.prototype.isTesselPort = function(port) {
  return Object.keys(this.ports).find(name => this.ports[name] === port) !== undefined;
};
```

```js
// In t1-firmware/builtin/tessel.js:
Tessel.prototype.isTesselPort = function(port) {
  for (var name in this.ports) {
    if (this.ports[name] === port) {
      return true
    }
  }
  return false;
};
```

Notes: 

- The t1-firmware version is different because [I don't trust `Object.keys(...)` in that runtime](https://github.com/tessel/t1-runtime/issues/126) and I don't know if `Array.prototype.find` is available there.
- Neither was actually tested



... Then in the module code, you can do this: 


```js
// Needs to be added also
var tessel = require('tessel');
```


```js
function use(hardware, callback) {
  if (!tessel.isTesselPort(hardware)) {
    console.warn(new Error('Improperly specified Tessel port.'));
  }

  return new Ambient(hardware, callback);
}
```

**Comment by rwaldron** [comment]
That also allows pre-populating the `tessel.ports` object with stubs/mocks

**Comment by wyze** [comment]
Awesome @rwaldron! I like that a lot. Do you have any thoughts around 1 & 2 from the original issue?

**Comment by rwaldron** [comment]
>  thoughts around 1 & 2 from the original issue

>> Why console.warn instead of throw? If it does warn, it is going to throw somewhere else, we should give feedback as early as possible.

I'm a fan of "fail early and hard", but I think that ship has sailed here? You could update it to throw, and if anyone ever files an issue, dial it back. WDYT?

>> Can this check be moved into the Ambient function to also catch rare cases where someone is not using the use function and just Ambient directly?

I think that makes sense.

**Comment by wyze** [comment]
> I'm a fan of "fail early and hard", but I think that ship has sailed here? You could update it to throw, and if anyone ever files an issue, dial it back. WDYT?

I think that would work. I think if they see this warning, they are going to run into other issues with not being able to communicate with the port.

**Comment by rwaldron** [comment]
Agreed. It's unlikely that this will cause any real problems in practice.

**Comment by rwaldron** [comment]
Something that just occurred to me: since you'll need to add `require('tessel');`, you should probably do so through an intermediate module. 

Create `test/common/tessel-mock.js`: 
```js
module.exports = {
  isTesselPort: function() {} // you'll stub the return in your tests
  // ...and other stuff...
};
```

Then in `index.js`:

```js
var tessel = global.IS_TEST_ENV ? require('../test/common/tessel-mock.js') : require('tessel');
```
